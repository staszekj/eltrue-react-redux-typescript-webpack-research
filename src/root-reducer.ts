import {combineReducers} from 'redux';
import {routerReducer as router, RouterState} from 'react-router-redux';

import {CounterState as CountersState, reducer as counters} from './counters/counter.reducer';
import {RootAction} from './root-action';

interface StoreEnhancerState {
}

export interface RootState extends StoreEnhancerState {
  router: RouterState;
  counters: CountersState;
}

export const rootReducer = combineReducers<RootState, RootAction>({
  router,
  counters,
});
