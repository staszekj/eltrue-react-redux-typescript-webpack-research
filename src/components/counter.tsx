import * as React from 'react';
import {ChangeEvent} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import {countersActions} from '../features/counters';
import {RootState} from '../features/root-reducer';
import {DispatchProps, OwnProps, OwnState, StateProps} from './types';

type CounterProps = OwnProps & StateProps & DispatchProps;

class CounterClazz extends React.Component<CounterProps, OwnState> {
  interval: number;
  state: OwnState = {count: 0};

  constructor(props: CounterProps) {
    super(props);
    this.onCounterTxtChange = this.onCounterTxtChange.bind(this);
  }

  onCounterTxtChange(event: ChangeEvent<HTMLInputElement>) {
    this.setState({...this.state, txt: event.target.value});
  }

  componentWillMount() {
    const incrementCounter = () => {
      this.setState({count: this.state.count + 1});
      this.props.reduxAdd(10);
    };
    this.interval = setInterval(incrementCounter, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    return (
      <div>
        <div>Counter: {this.state.count}</div>
        <div>Text: {this.state.txt}</div>
        <div>Redux Counter: {this.props.reduxCounter}</div>
        <div><input onChange={this.onCounterTxtChange}/></div>
      </div>
    );
  }
}

const mapStateToProps = (state: RootState): StateProps => {
  return {reduxCounter: state.counters.reduxCounter};
};

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  reduxAdd: bindActionCreators(countersActions.add, dispatch),
});

export const Counter = connect<StateProps, DispatchProps, OwnProps, RootState>
(mapStateToProps, mapDispatchToProps)(CounterClazz);
