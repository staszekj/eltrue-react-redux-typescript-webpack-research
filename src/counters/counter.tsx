import * as React from 'react';
import {ChangeEvent} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import * as countersActions from './counter.actions';
import {RootState} from '../root-reducer';
import {DispatchProps, OwnProps, OwnState, StateProps} from './counter.types';
import Display from '../7segments';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import RaisedButton from 'material-ui/RaisedButton';
import style from './counter.scss';

type CounterProps = OwnProps & StateProps & DispatchProps;

class CounterClazz extends React.Component<CounterProps, OwnState> {
  interval: number;
  state: OwnState = {count: 0, txt: ''};

  constructor(props: CounterProps) {
    super(props);
    this.onCounterTxtChange = this.onCounterTxtChange.bind(this);
    this.incrementBy100 = this.incrementBy100.bind(this);
  }

  onCounterTxtChange(event: ChangeEvent<HTMLInputElement>) {
    this.setState({...this.state, txt: event.target.value});
  }

  componentWillMount() {
    const incrementCounter = () => {
      this.setState({count: this.state.count + 1});
      this.props.reduxAdd(10);
    };
    this.interval = window.setInterval(incrementCounter, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  incrementBy100() {
    this.props.reduxAdd(100);
  }

  render() {
    return (
      <div>
        <div className={style.counterBlue}>Counter: {this.state.count}</div>
        <div>Text: {this.state.txt}</div>
        <div>Redux Counter: {this.props.reduxCounter}</div>
        <div>
          <MuiThemeProvider>
            <RaisedButton label="Add 100" onClick={this.incrementBy100}/>
          </MuiThemeProvider>
        </div>
        <Display value={this.props.reduxCounter}/>
        <div><input onChange={this.onCounterTxtChange}/></div>
      </div>
    );
  }
}

const mapStateToProps = (state: RootState): StateProps => {
  return {reduxCounter: state.counters.reduxCounter};
};

const mapDispatchToProps = (dispatch: Dispatch<RootState>): DispatchProps => ({
  reduxAdd: bindActionCreators(countersActions.add, dispatch),
});

export const Counter = connect<StateProps, DispatchProps, OwnProps, RootState>
(mapStateToProps, mapDispatchToProps)(CounterClazz);
