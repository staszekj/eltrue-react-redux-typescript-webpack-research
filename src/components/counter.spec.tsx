import * as React from 'react';
import {shallow} from 'enzyme';
import {Counter} from './counter';
import {Provider} from 'react-redux';
import {rootReducer} from '../features/root-reducer';
import {createStore} from 'redux';

describe('Counter', () => {
  const store = createStore(rootReducer, {});
  const component = shallow(<Provider store={store}><Counter/></Provider>);

  it('should match a snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should initialize state', () => {
    expect(component.state()).toEqual({count: 0});
  });

  it('should change state after a second', (done) => {
    setTimeout(() => {
      expect(component.state()).not.toEqual({count: 0});
      done();
    }, 1500);
  });
});
