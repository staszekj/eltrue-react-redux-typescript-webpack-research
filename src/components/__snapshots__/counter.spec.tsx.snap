// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should match a snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CounterClazz
    reduxAdd={[Function]}
    reduxCounter={0}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          Counter: 
          0
        </div>,
        <div>
          Text: 
        </div>,
        <div>
          Redux Counter: 
          0
        </div>,
        <div>
          <input
            onChange={[Function]}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Counter: ",
            0,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Counter: ",
          0,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Text: ",
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Text: ",
          undefined,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Redux Counter: ",
            0,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Redux Counter: ",
          0,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <input
            onChange={[Function]}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            Counter: 
            0
          </div>,
          <div>
            Text: 
          </div>,
          <div>
            Redux Counter: 
            0
          </div>,
          <div>
            <input
              onChange={[Function]}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Counter: ",
              0,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Counter: ",
            0,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Text: ",
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Text: ",
            undefined,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Redux Counter: ",
              0,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Redux Counter: ",
            0,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              onChange={[Function]}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
